Z3_get_bool_value can return UNDEF

lucasccordeiro opened the pull request:
Check whether the return value of `Z3_get_bool_value` can return UNDEF. In particular, according to the Z3 API, there are three possible outcomes for `Z3_get_bool_value`: `TRUE`, `FALSE`, and `UNDEF`. 

```
    /**
       \brief Return \c Z3_L_TRUE if \c a is true, \c Z3_L_FALSE if it is false, and \c Z3_L_UNDEF otherwise.

       def_API('Z3_get_bool_value', INT, (_in(CONTEXT), _in(AST)))
    */
    Z3_lbool Z3_API Z3_get_bool_value(Z3_context c, Z3_ast a);
```

We should abort if that function returns `UNDEF`, as we do for the other solvers.

lucasccordeiro committed:
Z3_get_bool_value can return UNDEF

Check the Z3_get_bool_value return value. There are three possible
outcomes: TRUE, FALSE, and UNDEF. We should abort if that function
returns UNDEF.

lucasccordeiro commented:
Our implementation for the other SMT solvers seem to support the method `get_bool` correctly:

```
boolector_convt::get_bool
cvc_convt::get_bool
mathsat_convt::get_bool
yices_convt::get_bool
```

rafaelsamenezes commented on src/solvers/z3/z3_conv.cpp:
```diff
@@ -1144,10 +1144,22 @@ bool z3_convt::get_bool(smt_astt a)
+    res = false;
+    break;
+  default:
+    abort();
```
Add an `std::cerr` message explaining why the abort ocurred

lucasccordeiro replied:
Done.

lucasccordeiro committed:
added error message for Z3 and Boolector

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented:
Do you have a test case where it fails?

Please check the documentation of model.eval for cases where can fail.

The easier fix is setting model completion to true in the eval method: it
will assign a value to the boolean, even if none is present in the model.

On Sun, 10 Jan 2021, 09:45 Lucas Cordeiro, <notifications@github.com> wrote:

> *@lucasccordeiro* commented on this pull request.
> ------------------------------
>
> In src/solvers/z3/z3_conv.cpp
> <https://github.com/esbmc/esbmc/pull/336#discussion_r554569827>:
>
> >
> -  return false;
> +  bool res;
> +  switch(result)
> +  {
> +  case Z3_L_TRUE:
> +    res = true;
> +    break;
> +  case Z3_L_FALSE:
> +    res = false;
> +    break;
> +  default:
> +    abort();
>
> Done.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/336#discussion_r554569827>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZLUUWSQKTZL4DNQVDSZGVO7ANCNFSM4V4NKQSQ>
> .
>

lucasccordeiro commented:
`01_cbmc_Fixedbv23` and`01_cbmc_Fixedbv24` are regression tests that fail:

```
esbmc main.c --fixedbv --no-slice --ir --z3 
....
Unwinding loop 2 iteration 1 file main.c line 25 function powInt
Unwinding loop 4 iteration 1 file main.c line 100 function iirOutFixed
Unwinding loop 1 iteration 1 file main.c line 9 function shiftL
Unwinding loop 10 iteration 6 file main.c line 156 function main
Symex completed in: 0.121s (1201 assignments)
Slicing time: 0.000s (removed 4 assignments)
Generated 218 VCC(s), 75 remaining after simplification (1197 assignments)
Encoding remaining VCC(s) using integer/real arithmetic
Encoding to solver time: 0.014s
Solving with solver Z3 v4.8.9
Encoding to solver time: 0.014s
Runtime decision procedure: 0.015s
Building error trace
Can't get boolean value from Z3
```

`test_20_flex`, `github_213_mutex_destroy_fail`, and `github_287_2` also fail:

```
esbmc main.c --z3
lucascordeiro@Lucass-MBP github_287_2 % ~/esbmc/build/src/esbmc/esbmc main.c --z3
...
Symex completed in: 0.000s (17 assignments)
Slicing time: 0.000s (removed 11 assignments)
Generated 3 VCC(s), 3 remaining after simplification (6 assignments)
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding to solver time: 0.000s
Solving with solver Z3 v4.8.9
Encoding to solver time: 0.000s
Runtime decision procedure: 0.001s
Building error trace
Can't get boolean value from Z3
```

lucasccordeiro committed:
set model completion to TRUE for constant Booleans

Z3 will assign an interpretation to the Boolean constants,
which are essentially don't cares.

lucasccordeiro committed:
enabled test cases after setting completion model

This commit enables various test cases, which now
produce some useful result when completion model
is enabled.

lucasccordeiro commented:
@mikhailramalho: After enabling `completion_model` to TRUE, we could enable various regression tests that were broken.

What do you think if we also set `completion true` to TRUE in:

```
ieee_floatt z3_convt::get_fpbv(smt_astt a)
```

and

```
z3_convt::get_array_elem(smt_astt array, uint64_t index, const type2tc &subtype)
```

I suspect that it will have a positive impact for FuSeBMC for test case generation.

mikhailramalho commented:
I can't reproduce the issue here, and it doesn't seem to be triggering our
CI.



Em dom., 10 de jan. de 2021 às 12:54, Lucas Cordeiro <
notifications@github.com> escreveu:

> 01_cbmc_Fixedbv23 and01_cbmc_Fixedbv24 are regression tests that fail:
>
> esbmc main.c --fixedbv --no-slice --ir --z3
> ....
> Unwinding loop 2 iteration 1 file main.c line 25 function powInt
> Unwinding loop 4 iteration 1 file main.c line 100 function iirOutFixed
> Unwinding loop 1 iteration 1 file main.c line 9 function shiftL
> Unwinding loop 10 iteration 6 file main.c line 156 function main
> Symex completed in: 0.121s (1201 assignments)
> Slicing time: 0.000s (removed 4 assignments)
> Generated 218 VCC(s), 75 remaining after simplification (1197 assignments)
> Encoding remaining VCC(s) using integer/real arithmetic
> Encoding to solver time: 0.014s
> Solving with solver Z3 v4.8.9
> Encoding to solver time: 0.014s
> Runtime decision procedure: 0.015s
> Building error trace
> Can't get boolean value from Z3
>
> test_20_flex, github_213_mutex_destroy_fail, and github_287_2 also fail:
>
> esbmc main.c --z3
> lucascordeiro@Lucass-MBP github_287_2 % ~/esbmc/build/src/esbmc/esbmc main.c --z3
> ...
> Symex completed in: 0.000s (17 assignments)
> Slicing time: 0.000s (removed 11 assignments)
> Generated 3 VCC(s), 3 remaining after simplification (6 assignments)
> Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
> Encoding to solver time: 0.000s
> Solving with solver Z3 v4.8.9
> Encoding to solver time: 0.000s
> Runtime decision procedure: 0.001s
> Building error trace
> Can't get boolean value from Z3
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/336#issuecomment-757507373>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH5P6MKVRG33TACG3ZLSZHLSXANCNFSM4V4NKQSQ>
> .
>


-- 

Mikhail Ramalho.

lucasccordeiro commented:
I got these failing tests from CI: https://github.com/esbmc/esbmc/pull/336/checks?check_run_id=1676818423

lucasccordeiro commented:
Here is the link to visualise them in our CI:

https://github.com/esbmc/esbmc/pull/336/checks?check_run_id=1676818423

Em dom., 10 de jan. de 2021 às 17:12, Mikhail R. Gadelha <
notifications@github.com> escreveu:

> I can't reproduce the issue here, and it doesn't seem to be triggering our
> CI.
>
>
>
> Em dom., 10 de jan. de 2021 às 12:54, Lucas Cordeiro <
> notifications@github.com> escreveu:
>
> > 01_cbmc_Fixedbv23 and01_cbmc_Fixedbv24 are regression tests that fail:
> >
> > esbmc main.c --fixedbv --no-slice --ir --z3
> > ....
> > Unwinding loop 2 iteration 1 file main.c line 25 function powInt
> > Unwinding loop 4 iteration 1 file main.c line 100 function iirOutFixed
> > Unwinding loop 1 iteration 1 file main.c line 9 function shiftL
> > Unwinding loop 10 iteration 6 file main.c line 156 function main
> > Symex completed in: 0.121s (1201 assignments)
> > Slicing time: 0.000s (removed 4 assignments)
> > Generated 218 VCC(s), 75 remaining after simplification (1197
> assignments)
> > Encoding remaining VCC(s) using integer/real arithmetic
> > Encoding to solver time: 0.014s
> > Solving with solver Z3 v4.8.9
> > Encoding to solver time: 0.014s
> > Runtime decision procedure: 0.015s
> > Building error trace
> > Can't get boolean value from Z3
> >
> > test_20_flex, github_213_mutex_destroy_fail, and github_287_2 also fail:
> >
> > esbmc main.c --z3
> > lucascordeiro@Lucass-MBP github_287_2 % ~/esbmc/build/src/esbmc/esbmc
> main.c --z3
> > ...
> > Symex completed in: 0.000s (17 assignments)
> > Slicing time: 0.000s (removed 11 assignments)
> > Generated 3 VCC(s), 3 remaining after simplification (6 assignments)
> > Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
> > Encoding to solver time: 0.000s
> > Solving with solver Z3 v4.8.9
> > Encoding to solver time: 0.000s
> > Runtime decision procedure: 0.001s
> > Building error trace
> > Can't get boolean value from Z3
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/esbmc/esbmc/pull/336#issuecomment-757507373>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAKEJH5P6MKVRG33TACG3ZLSZHLSXANCNFSM4V4NKQSQ
> >
> > .
> >
>
>
> --
>
> Mikhail Ramalho.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/336#issuecomment-757510170>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4F4HJXWZFBHBVZBCLH5I3SZHNZDANCNFSM4V4NKQSQ>
> .
>


-- 
Lucas C. Cordeiro​

https://ssvlab.github.io/lucasccordeiro/

mikhailramalho commented:
Sure, let's set the model completion to true for them all.

Em dom., 10 de jan. de 2021 às 13:17, Lucas Cordeiro <
notifications@github.com> escreveu:

> Here is the link to visualise them in our CI:
>
> https://github.com/esbmc/esbmc/pull/336/checks?check_run_id=1676818423
>
> Em dom., 10 de jan. de 2021 às 17:12, Mikhail R. Gadelha <
> notifications@github.com> escreveu:
>
> > I can't reproduce the issue here, and it doesn't seem to be triggering
> our
> > CI.
> >
> >
> >
> > Em dom., 10 de jan. de 2021 às 12:54, Lucas Cordeiro <
> > notifications@github.com> escreveu:
> >
> > > 01_cbmc_Fixedbv23 and01_cbmc_Fixedbv24 are regression tests that fail:
> > >
> > > esbmc main.c --fixedbv --no-slice --ir --z3
> > > ....
> > > Unwinding loop 2 iteration 1 file main.c line 25 function powInt
> > > Unwinding loop 4 iteration 1 file main.c line 100 function iirOutFixed
> > > Unwinding loop 1 iteration 1 file main.c line 9 function shiftL
> > > Unwinding loop 10 iteration 6 file main.c line 156 function main
> > > Symex completed in: 0.121s (1201 assignments)
> > > Slicing time: 0.000s (removed 4 assignments)
> > > Generated 218 VCC(s), 75 remaining after simplification (1197
> > assignments)
> > > Encoding remaining VCC(s) using integer/real arithmetic
> > > Encoding to solver time: 0.014s
> > > Solving with solver Z3 v4.8.9
> > > Encoding to solver time: 0.014s
> > > Runtime decision procedure: 0.015s
> > > Building error trace
> > > Can't get boolean value from Z3
> > >
> > > test_20_flex, github_213_mutex_destroy_fail, and github_287_2 also
> fail:
> > >
> > > esbmc main.c --z3
> > > lucascordeiro@Lucass-MBP github_287_2 % ~/esbmc/build/src/esbmc/esbmc
> > main.c --z3
> > > ...
> > > Symex completed in: 0.000s (17 assignments)
> > > Slicing time: 0.000s (removed 11 assignments)
> > > Generated 3 VCC(s), 3 remaining after simplification (6 assignments)
> > > Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
> > > Encoding to solver time: 0.000s
> > > Solving with solver Z3 v4.8.9
> > > Encoding to solver time: 0.000s
> > > Runtime decision procedure: 0.001s
> > > Building error trace
> > > Can't get boolean value from Z3
> > >
> > > —
> > > You are receiving this because your review was requested.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/esbmc/esbmc/pull/336#issuecomment-757507373>, or
> > > unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAKEJH5P6MKVRG33TACG3ZLSZHLSXANCNFSM4V4NKQSQ
> > >
> > > .
> > >
> >
> >
> > --
> >
> > Mikhail Ramalho.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/esbmc/esbmc/pull/336#issuecomment-757510170>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AA4F4HJXWZFBHBVZBCLH5I3SZHNZDANCNFSM4V4NKQSQ
> >
> > .
> >
>
>
> --
> Lucas C. Cordeiro​
>
> https://ssvlab.github.io/lucasccordeiro/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/336#issuecomment-757510880>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZD4PWOJJQ2XACK3XLSZHOLXANCNFSM4V4NKQSQ>
> .
>


-- 

Mikhail Ramalho.

lucasccordeiro committed:
set completion model to true for FP and arrays

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

rafaelsamenezes reviewed (approved):
LGTM

lucasccordeiro commented:
We're now setting the model completion to TRUE; with this modification, we were able to enable three regression tests. We have also improved the error message provided by Z3 and Boolector.

I'll merge this PR for now.
